src/GafferScene/SetAlgo.cpp

```diff
@@ -0,0 +1,455 @@
+namespace ascii = boost::spirit::ascii;
+namespace phx = boost::phoenix;
+
+struct BinaryOp;
+struct nil {};
+
+// Wrap a string into a SetName struct that gives it semantic meaning
+struct SetName
+{
+	std::string name;
+};
+
+BOOST_FUSION_ADAPT_STRUCT(
+	SetName,
+	(std::string, name)
+)
+
+// Wrap a string into an ObjectName struct that gives it semantic meaning
+struct ObjectName
+{
+	std::string name;
+};
+
+BOOST_FUSION_ADAPT_STRUCT(
+ ObjectName,
+ (std::string, name)
+)
+
+// Determine which Ops are supported in SetExpressions
+// and provide a way to print them for debugging.
+enum Op { And, Or, AndNot };
+
+std::ostream & operator<<( std::ostream &out, const Op &op )
+{
+	switch( op )
+	{
+	case Or :
+		out << "|"; break;
+	case And :
+		out << "&"; break;
+	case AndNot :
+		out << "-"; break;
+	}
+	return out;
+}
+
+struct ExpressionAst
+{
+	typedef
+	boost::variant<
+		nil, // can't happen!
+		SetName,
+		ObjectName,
+		boost::recursive_wrapper<ExpressionAst>,
+		boost::recursive_wrapper<BinaryOp>
+		>
+	type;
+
+	ExpressionAst()
+		: expr( nil() ) {}
+
+	template <typename Expr>
+	ExpressionAst( Expr const& expr )
+		: expr( expr ) {}
+
+	ExpressionAst& operator+=( ExpressionAst const& rhs );
+	ExpressionAst& operator-=( ExpressionAst const& rhs );
+	ExpressionAst& operator/=( ExpressionAst const& rhs );
+
+	type expr;
+};
+
+struct BinaryOp
+{
+	BinaryOp(
+		ExpressionAst const& left,
+		Op op,
+		ExpressionAst const& right )
+		: left( left ), op(op), right( right ) {}
+
+	ExpressionAst left;
+	Op op;
+	ExpressionAst right;
+};
+
+ExpressionAst& ExpressionAst::operator+=( ExpressionAst const& rhs )
+{
+	expr = BinaryOp( expr, And, rhs );
+	return *this;
+}
+
+ExpressionAst& ExpressionAst::operator-=( ExpressionAst const& rhs )
+{
+	expr = BinaryOp( expr, AndNot , rhs );
+	return *this;
+}
+
+ExpressionAst& ExpressionAst::operator/=( ExpressionAst const& rhs )
+{
+	expr = BinaryOp( expr, Or , rhs );
+	return *this;
+}
+
+// Visiting the AST
+// ----------------
+// For a simple AST with only one operation (AND) on two sets (A and B)
+// the output will look like this: op:&(A, B)
+// If one of the operands is an operation itself: op:&(A, op:|(B, C))
+struct AstPrinter
+{
+	typedef void result_type;
+
+	void operator()( std::string n ) const { std::cout << n; }
+	void operator()( ObjectName n ) const { std::cout << n.name; }
+	void operator()( SetName n ) const { std::cout << n.name; }
+
+	void operator()( ExpressionAst const& ast ) const
+	{
+		boost::apply_visitor( *this, ast.expr );
+	}
+
+	void operator()( BinaryOp const& expr ) const
+	{
+		std::cout << "op:" << expr.op << "(";
+		boost::apply_visitor( *this, expr.left.expr );
+		std::cout << ", ";
+		boost::apply_visitor( *this, expr.right.expr );
+		std::cout << ')';
+	}
+};
+
+// support for debugging through BOOST_SPIRIT_DEBUG
+std::ostream& operator << (std::ostream& stream, const ExpressionAst& expr)
+{
+	AstPrinter printer;
+	printer(expr);
```

Comment:
This is a little messy, I guess, because it's not guaranteed that the `AstPrinter` instance writes into `stream`. We could add a constructor for `AstPrinter` that stores a given `std::ostream` reference, but I'm not sure it's worth it for simple debugging stuff.